larry/backup.go

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2025 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package larry
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/dustin/go-humanize"
+)
+
+// DumpTables and Restore are used to create portable backups. Any
+// Encoder/Decoder can be used provided it conforms to the interface.  The
+// caller is responsible for quiescing the tables. The backup itself will be
+// made using a consistent view.
+//
+// Typical use:
+//	var b bytes.Buffer
+//	zw, _ := zstd.NewWriter(&b)
+//	je := json.NewEncoder(zw)
+//	_ = db.DumpTables(ctx, tables, je)
+//	_ = zw.Close()
+//
+//	zr, _ := zstd.NewReader(&b)
+//	jd := json.NewDecoder(zr)
+//	_ = db.Restore(ctx, jd)
+
+// OperationT is used to tag what database operation is being described. These
+// operations are used to perform database dump/restores.
+type OperationT uint8
+
+const (
+	OpPut OperationT = 1 // Database put
+	OpDel OperationT = 2 // Database delete
+)
+
+// Operation describes a single database operation. An entire database can be
+// replicated by creating a delta between a source database table and a
+// destination database table.
+type Operation struct {
+	Op    OperationT `json:"op"`
+	Table string     `json:"table"`
+	Key   []byte     `json:"key"`
+	Value []byte     `json:"value,omitempty" binary:"omitempty"`
+}
+
+// Replay describes a replay of operations into a database.
+type Replay struct {
+	Operations []Operation `json:"operations"`
+}
+
+type Encoder interface {
+	Encode(v any) error
+}
+
+type Decoder interface {
+	Decode(v any) error
+}
+
+var (
+	DefaultMaxRestoreChunk = 256 * 1024 * 1024 // Maximum transaction size during restore.
+
+	Verbose = true // use logger to print progress
+)
+
+// DumpTables is used to create a portable backup of a set of database tables.
+func DumpTables(ctx context.Context, db Database, tables []string, target Encoder) error {
+	for _, table := range tables {
+		it, err := db.NewIterator(ctx, table)
+		if err != nil {
+			return err
+		}
+		for it.Next(ctx) {
+			op := Operation{
+				Op:    OpPut,
+				Table: table,
+				Key:   it.Key(ctx),
+				Value: it.Value(ctx),
+			}
+			err := target.Encode(op)
+			if err != nil {
+				it.Close(ctx)
+				return err
+			}
+		}
+		it.Close(ctx)
+	}
+
+	return nil
+}
+
+// Restore is used to replay a portable backup of a database.
+func Restore(ctx context.Context, db Database, source Decoder) error {
+	batch, err := db.NewBatch(ctx)
+	if err != nil {
+		return err
+	}
+
+	var done bool
+	for !done {
+		totalWritten := 0
```

Comment:
nit: can use `var` with empty value
```suggestion
		var totalWritten int
```

Reply:
That would work but I prefer `:=` in this specific case to show that it is set to `0` every iteration.  Using `var` is conusing in this case because it could seem like it was initiated and it would not be reset.